src/argon2_opt_plug.c

```diff
@@ -228,36 +230,47 @@ void argon2_fill_segment(const argon2_instance_t *instance,
     curr_offset = position.lane * instance->lane_length +
                   position.slice * instance->segment_length + starting_index;
 
-    if (0 == curr_offset % instance->lane_length) {
+    if ((0 == position.slice) && (0 == starting_index)) {
         /* Last block in this lane */
         prev_offset = curr_offset + instance->lane_length - 1;
     } else {
         /* Previous block */
         prev_offset = curr_offset - 1;
     }
 
+    /* Fixed point multiply constant for dividing by lanes */
+    if ((lanes & (lanes - 1)) != 0) {
+        lanes_reciprocal = (uint32_t) (UINT64_C(0x100000000) / lanes);
+    }
+
     memcpy(state, ((instance->memory + prev_offset)->v), ARGON2_BLOCK_SIZE);
 
+    i_address = starting_index;
     for (i = starting_index; i < instance->segment_length;
-         ++i, ++curr_offset, ++prev_offset) {
-        /*1.1 Rotating prev_offset if needed */
-        if (curr_offset % instance->lane_length == 1) {
-            prev_offset = curr_offset - 1;
-        }
-
+         ++i, ++curr_offset) {
         /* 1.2 Computing the index of the reference block */
         /* 1.2.1 Taking pseudo-random value from the previous block */
         if (data_independent_addressing) {
-            if (i % ARGON2_ADDRESSES_IN_BLOCK == 0) {
+            if (i_address == 0) {
```

Comment:
I'm not sure we want this set of changes for `% ARGON2_ADDRESSES_IN_BLOCK`, which was likely just an AND. I suggest looking at the generated code with/without these changes.